packages/chain-bridge/src/adapters/StargateAdapter.sol

```diff
@@ -221,9 +222,92 @@ contract StargateAdapter is
                 stargateContract
             );
 
-        // Get messaging fee and perform transfer
-        MessagingFee memory messagingFee = stargate.quoteSend(sendParam, false);
+        // Determine fee payment mode and get messaging fee
+        bool payInToken = options.payInProtocolToken &&
+            protocolFeeToken != address(0);
+        MessagingFee memory messagingFee = stargate.quoteSend(
+            sendParam,
+            payInToken
+        );
+
+        if (payInToken) {
+            _executeTokenPayment(
+                operationId,
+                params,
+                stargate,
+                sendParam,
+                messagingFee,
+                providedFee
+            );
+        } else {
+            _executeNativePayment(
+                params,
+                stargate,
+                sendParam,
+                messagingFee,
+                providedFee
+            );
+        }
+    }
 
+    /**
+     * @dev Execute token payment for Stargate transfer
+     * @param operationId The operation ID for this transfer
+     * @param params Transfer parameters
+     * @param stargate Stargate contract instance
+     * @param sendParam Send parameters for Stargate
+     * @param messagingFee Fee information from Stargate
+     * @param providedFee Native fee provided by caller
+     */
+    function _executeTokenPayment(
+        bytes32 operationId,
+        BridgeTypes.ExecuteTransferParams memory params,
+        IStargateV2 stargate,
+        SendParam memory sendParam,
+        MessagingFee memory messagingFee,
+        uint256 providedFee
+    ) internal {
+        // Enforce that native fee should be zero in token mode
+        if (messagingFee.nativeFee != 0)
+            revert InsufficientFee(0, messagingFee.nativeFee);
+
+        // Use base contract functionality for protocol token fee collection
+        _collectProtocolTokenFee(
+            operationId,
+            params.originator,
```

Comment:
There is no handling for the token fee approval / check in the originator - so neither FleetProxy nor CrossChainArk